Add WasmGC article (#7439)
GoogleChrome/developer.chrome.com: site/en/blog/wasmgc/index.md

description: >
  There are two types of programming languages: garbage-collected programming languages and programming languages that require manual memory management. This article explains how thanks to WebAssembly Garbage Collection, short WasmGC, garbage-collected languages can be ported to WebAssembly.
authors:
  - thomassteiner
date: 2023-10-31
# updated: 2023-10-31
hero: image/8WbTDNrhLsU0El80frMBGE4eMCD3/gc5UyHFm1C4tPAQC6D1o.png
alt: Pictogram of a person throwing something in a trash can.
tags:
  - wasm
---

There are two types of programming languages: garbage-collected programming languages and programming languages that require manual memory management. Examples of the former, among many more, are Kotlin, PHP, or Java. Examples of the latter are C, C++, or Rust. As a general rule, higher-level programming languages are more likely to have garbage collection as a standard feature. In this blog post, the focus is on such garbage-collected programming languages and how they can be compiled to WebAssembly (Wasm). But what is garbage collection (often referred to as GC) to begin with?

{% BrowserCompat 'webassembly.garbage-collection' %}

## Garbage collection

> In simplified terms, the idea of garbage collection is the attempt to reclaim memory which was allocated by the program, but that is no longer referenced. Such memory is called garbage. There are many strategies for implementing garbage collection. One of them is [reference counting](https://en.wikipedia.org/wiki/Reference_counting) where the objective is to count the number of references to objects in memory. When there are no more references to an object, it can be marked as no longer used and thus ready for garbage collection. [PHP](https://www.php.net/)'s garbage collector [uses  reference counting](https://www.php.net/manual/en/features.gc.refcounting-basics.php), and using the [Xdebug](https://xdebug.org/) extension's [`xdebug_debug_zval()``](https://xdebug.org/docs/all_functions#xdebug_debug_zval) function allows you to peek under its hood. Consider the following PHP program.

Reviewer: ```suggestion
In simplified terms, the idea of garbage collection is the attempt to reclaim memory which was allocated by the program, but that is no longer referenced. Such memory is called garbage. There are many strategies for implementing garbage collection. One of them is [reference counting](https://en.wikipedia.org/wiki/Reference_counting) where the objective is to count the number of references to objects in memory. When there are no more references to an object, it can be marked as no longer used and thus ready for garbage collection. [PHP](https://www.php.net/)'s garbage collector [uses  reference counting](https://www.php.net/manual/en/features.gc.refcounting-basics.php), and using the [Xdebug](https://xdebug.org/) extension's [`xdebug_debug_zval()`](https://xdebug.org/docs/all_functions#xdebug_debug_zval) function allows you to peek under its hood. Consider the following PHP program.
```

## Traditional methods of porting languages to the Wasm runtime

Independently from the platform PHP is running on, PHP scripts are compiled into the same bytecode and run by the [Zend Engine](https://en.wikipedia.org/wiki/Zend_Engine). The Zend Engine is a compiler and runtime environment for the PHP scripting language. It consists of the Zend Virtual Machine (VM), which is composed of the Zend Compiler and the Zend Executor. Languages like PHP that are implemented in other high-level languages like C commonly have optimizations that target specific architectures, such as Intel or ARM, and require a different backend for each architecture. In this context, Wasm represents a new architecture. If the VM has architecture-specific code, like just-in-time (JIT) or ahead-of-time (AOT) compilation, then the developer also implements a backend for JIT/AOT for the new architecture. This approach makes a lot of sense because often the main part of the codebase can just be recompiled for each new architecture.

> Given how low-level Wasm is, it is natural to try the same approach there: Recompile the main VM code with its parser, library support, garbage collection, and optimizer to Wasm, and implement a JIT or AOT backend for Wasm if needed. This has been possible since the Wasm MVP, and it works well in practice in many cases. In fact, [PHP compiled to Wasm](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/php-wasm/compile/build.js) is what powers the [WordPress Playground](https://playground.wordpress.net/). Learn more about the project in the article [Build in-browser WordPress experiences with WordPress Playground and WebAssembly](https://web.dev/wordpress-playground/).

Reviewer: ```suggestion

Given how low-level Wasm is, it is natural to try the same approach there: Recompile the main VM code with its parser, library support, garbage collection, and optimizer to Wasm, and implement a JIT or AOT backend for Wasm if needed. This has been possible since the Wasm MVP, and it works well in practice in many cases. In fact, [PHP compiled to Wasm](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/php-wasm/compile/build.js) is what powers the [WordPress Playground](https://playground.wordpress.net/). Learn more about the project in the article [Build in-browser WordPress experiences with WordPress Playground and WebAssembly](https://web.dev/wordpress-playground/).
```

> However, PHP Wasm runs in the browser in the context of the host language JavaScript. In Chrome, [JavaScript and Wasm are run in V8](https://v8.dev/docs/wasm-compilation-pipeline), Google's open source JavaScript engine that implements ECMAScript as specified in [ECMA-262](https://www.ecma-international.org/publications-and-standards/standards/ecma-262/). And, [V8 already has a garbage collector](https://chromium.googlesource.com/v8/v8.git/+/refs/heads/main/src/heap/cppgc/). This means developers making use of, for example, PHP compiled to Wasm, end up shipping a garbage collector implementation of the ported language (PHP) to the browser that already has a garbage collector, which is as wasteful as it sounds. This is where WasmGC comes in.

Reviewer: ```suggestion

However, PHP Wasm runs in the browser in the context of the host language JavaScript. In Chrome, [JavaScript and Wasm are run in V8](https://v8.dev/docs/wasm-compilation-pipeline), Google's open source JavaScript engine that implements ECMAScript as specified in [ECMA-262](https://www.ecma-international.org/publications-and-standards/standards/ecma-262/). And, [V8 already has a garbage collector](https://chromium.googlesource.com/v8/v8.git/+/refs/heads/main/src/heap/cppgc/). This means developers making use of, for example, PHP compiled to Wasm, end up shipping a garbage collector implementation of the ported language (PHP) to the browser that already has a garbage collector, which is as wasteful as it sounds. This is where WasmGC comes in.
```

> The other problem of the old approach of letting Wasm modules build their own GC on top of Wasm's linear memory is that there's then no interaction between Wasm's own garbage collector and the built-on-top garbage collector of the compiled-to-Wasm language, which tends to cause problems like memory leaks and inefficient collection attempts. Letting Wasm modules reuse the existing built-in GC avoids these issues.

Reviewer: ```suggestion

The other problem of the old approach of letting Wasm modules build their own GC on top of Wasm's linear memory is that there's then no interaction between Wasm's own garbage collector and the built-on-top garbage collector of the compiled-to-Wasm language, which tends to cause problems like memory leaks and inefficient collection attempts. Letting Wasm modules reuse the existing built-in GC avoids these issues.
```

> To verify the real-world impact of this improvement, Chrome's Wasm team has compiled versions of the [Fannkuch benchmark](https://benchmarksgame-team.pages.debian.net/benchmarksgame/description/fannkuchredux.html) (which allocates data structures as it works) from [C](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-gcc-5.html), [Rust](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-rust-2.html), and [Java](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-java-2.html). The C and Rust binaries could be anywhere from _6.1 K_ to _9.6 K_ depending on the various compiler flags, while the Java version is much smaller at only _2.3 K_! C and Rust do not include a garbage collector, but they do still bundle `malloc/free` to manage memory, and the reason Java is smaller here is because it doesn't need to bundle any memory management code at all. This is just one specific example, but it shows that WasmGC binaries have the potential of being very small, and this is even before any significant work on optimizing for size.

Reviewer: ```suggestion

To verify the real-world impact of this improvement, Chrome's Wasm team has compiled versions of the [Fannkuch benchmark](https://benchmarksgame-team.pages.debian.net/benchmarksgame/description/fannkuchredux.html) (which allocates data structures as it works) from [C](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-gcc-5.html), [Rust](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-rust-2.html), and [Java](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-java-2.html). The C and Rust binaries could be anywhere from _6.1 K_ to _9.6 K_ depending on the various compiler flags, while the Java version is much smaller at only _2.3 K_! C and Rust do not include a garbage collector, but they do still bundle `malloc/free` to manage memory, and the reason Java is smaller here is because it doesn't need to bundle any memory management code at all. This is just one specific example, but it shows that WasmGC binaries have the potential of being very small, and this is even before any significant work on optimizing for size.
```

---
layout: 'layouts/blog-post.njk'

> title: WebAssembly Garbage Collection (WasmGC)

Author: ```suggestion
title: WebAssembly Garbage Collection (WasmGC) now enabled by default in Chrome
```

To verify the real-world impact of this improvement, Chrome's Wasm team has compiled versions of the [Fannkuch benchmark](https://benchmarksgame-team.pages.debian.net/benchmarksgame/description/fannkuchredux.html) (which allocates data structures as it works) from [C](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-gcc-5.html), [Rust](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-rust-2.html), and [Java](https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-java-2.html). The C and Rust binaries could be anywhere from _6.1 K_ to _9.6 K_ depending on the various compiler flags, while the Java version is much smaller at only _2.3 K_! C and Rust do not include a garbage collector, but they do still bundle `malloc/free` to manage memory, and the reason Java is smaller here is because it doesn't need to bundle any memory management code at all. This is just one specific example, but it shows that WasmGC binaries have the potential of being very small, and this is even before any significant work on optimizing for size.

## Seeing a WasmGC-ported programming language in action

> One of the first programming languages that has been ported to Wasm thanks to WasmGC is [Kotlin](https://kotlinlang.org/) in the form of [Kotlin/Wasm](https://kotlinlang.org/docs/wasm-overview.html). The [demo](https://kotlin-wasm-hello-world.glitch.me/), with [source code](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/browser-example) courtesy of the Kotlin team, is shown in the following listing.

Author: ```suggestion
One of the first programming languages that has been ported to Wasm thanks to WasmGC is [Kotlin](https://kotlinlang.org/) in the form of [Kotlin/Wasm](https://kotl.in/wasmgc). The [demo](https://kotlin-wasm-hello-world.glitch.me/), with [source code](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/browser-example) courtesy of the Kotlin team, is shown in the following listing.
```

> One of the first programming languages that has been ported to Wasm thanks to WasmGC is [Kotlin](https://kotlinlang.org/) in the form of [Kotlin/Wasm](https://kotlinlang.org/docs/wasm-overview.html). The [demo](https://kotlin-wasm-hello-world.glitch.me/), with [source code](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/browser-example) courtesy of the Kotlin team, is shown in the following listing.

Author: Done in https://github.com/GoogleChrome/developer.chrome.com/pull/7439/commits/a77f47a7610f1b4ca0147b1bc2b259330e00eb0c.

> One of the first programming languages that has been ported to Wasm thanks to WasmGC is [Kotlin](https://kotlinlang.org/) in the form of [Kotlin/Wasm](https://kotl.in/wasmgc). The [demo](https://kotlin-wasm-hello-world.glitch.me/), with [source code](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/browser-example) courtesy of the Kotlin team, is shown in the following listing.

Reviewer 2: @tomayac please pull recent changes for examples and republish your examples before publishing the post.

I added more browser API usages to the simple browser example and made it a bit more interactive.
Also, other examples, including image viewer, got some cosmetic changes.

More importantly, all examples use the latest Kotlin toolchain, which generates final WasmGC opcodes.

Author: Done, both samples were updated to the latest version.

Reviewer 2: Great, thanks!

```kotlin
import kotlinx.browser.document
import kotlinx.dom.appendText
import org.w3c.dom.HTMLDivElement

fun main() {
    (document.getElementById("warning") as HTMLDivElement).style.display = "none"
    document.body?.appendText("Hello, ${greet()}!")
}

fun greet() = "world"
```

> {% Glitch id="kotlin-wasm-hello-world" %}

Author: ```suggestion
{% Glitch id="kotlin-wasm-hello-world", height="600" %}
```

{% Aside %}
To see the demo, you need to set the browser flags as per the instructions on the screen.
{% endAside %}

Now you may be wondering what the point is, since the Kotlin code above basically consists of the [JavaScript OM APIs converted to Kotlin](https://kotlinlang.org/api/latest/jvm/stdlib/org.w3c.dom/). It starts to make more sense in combination with [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/), which allows developers to build upon the UI they may already have created for their Android Kotlin apps. Check out an early exploration of this with the [Kotlin/Wasm image viewer](https://kotlin-wasm-image-viewer.glitch.me/) demo and explore its [source code](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/compose-imageviewer), likewise courtesy of the Kotlin team.

> {% Glitch id="kotlin-wasm-image-viewer" %}

Author: ```suggestion
{% Glitch id="kotlin-wasm-image-viewer", height="900" %}
```

{% Aside 'warning' %}

> The [Kotlin/Wasm image viewer](https://kotlin-wasm-image-viewer.glitch.me/) demo is fully rendered onto a `canvas` with no meaningful DOM tree that could be used to create an [accessibility tree](https://developer.mozilla.org/docs/Glossary/Accessibility_tree). This means it's completely inaccessible to non-sighted users and breaks important browser features like find on page, translation, text selection, extensions, zooming, and link to text fragment.

Author: ```suggestion
In this early experimental stage, the [Kotlin/Wasm image viewer](https://kotlin-wasm-image-viewer.glitch.me/) demo is fully rendered onto a `canvas` with no meaningful DOM tree that could be used to create an [accessibility tree](https://developer.mozilla.org/docs/Glossary/Accessibility_tree). This means it's completely inaccessible to non-sighted users and breaks important browser features like find on page, translation, text selection, extensions, zooming, and link to text fragment.
```

{% endAside %}

## Learn more about WasmGC

This blog post has barely scratched the surface and mostly provided a high-level overview of WasmGC. To learn more about the feature, check out these links:

> - [Wasm GC blogpost](https://docs.google.com/document/d/1zF0hh3PobeErcGTf2v2ppkbsKFHCtsXNP3VRMQO5ez8/edit)

Author: ```suggestion
- [Wasm GC blogpost](https://docs.google.com/document/d/1zF0hh3PobeErcGTf2v2ppkbsKFHCtsXNP3VRMQO5ez8/edit)
```
```suggestion
<!-- - [Wasm GC blogpost](https://docs.google.com/document/d/1zF0hh3PobeErcGTf2v2ppkbsKFHCtsXNP3VRMQO5ez8/edit) -->
```
